presto-cache/src/main/java/com/facebook/presto/cache/LocalRangeCacheManager.java

```diff
@@ -0,0 +1,461 @@
+
+            this.cacheRemovalExecutor.submit(() -> Arrays.stream(files).forEach(file -> {
+                try {
+                    Files.delete(file.toPath());
+                }
+                catch (IOException e) {
+                    // ignore
+                }
+            }));
+        }
+    }
+
+    @PreDestroy
+    public void destroy()
+    {
+        cacheFlushExecutor.shutdownNow();
+        cacheRemovalExecutor.shutdownNow();
+    }
+
+    @Override
+    public boolean get(FileCacheKey key, byte[] buffer, int offset)
+    {
+        boolean result = read(key, buffer, offset);
+        if (result) {
+            stats.incrementCacheHit();
+        }
+        else {
+            stats.incrementCacheMiss();
+        }
+
+        return result;
+    }
+
+    @Override
+    public void put(FileCacheKey key, Slice data)
+    {
+        if (stats.getInMemoryRetainedBytes() + data.length() >= maxInflightBytes) {
+            // cannot accept more requests
+            return;
+        }
+
+        // make a copy given the input data could be a reusable buffer
+        stats.addInMemoryRetainedBytes(data.length());
+        byte[] copy = data.getBytes();
+        cacheFlushExecutor.submit(() -> {
+            Path newPathHint = new Path(directory.toUri() + "/" + randomUUID() + EXTENSION);
+            if (!write(key, copy, newPathHint)) {
+                log.warn("%s Fail to persist cache %s with length %s ", Thread.currentThread().getName(), newPathHint, key.getLength());
+            }
+            stats.addInMemoryRetainedBytes(-copy.length);
+        });
+    }
+
+    private boolean read(FileCacheKey key, byte[] buffer, int offset)
+    {
+        if (key.getLength() <= 0) {
+            // no-op
+            return true;
+        }
+
+        try {
+            cache.get(key.getPath(), () -> true);
+        }
+        catch (ExecutionException e) {
+            // ignore
+        }
+
+        // check if the file is cached on local disk
+        CacheRange cacheRange = persisted.get(key.getPath());
+        if (cacheRange == null) {
+            return false;
+        }
+
+        LocalCacheFile cacheFile;
+        Lock readLock = cacheRange.getLock().readLock();
+        readLock.lock();
+        try {
+            Map<Range<Long>, LocalCacheFile> diskRanges = cacheRange.getRange().subRangeMap(Range.closedOpen(key.getOffset(), key.getLength() + key.getOffset())).asMapOfRanges();
+            if (diskRanges.size() != 1) {
+                // no range or there is a hole in between
+                return false;
+            }
+
+            cacheFile = getOnlyElement(diskRanges.entrySet().iterator()).getValue();
+        }
+        finally {
+            readLock.unlock();
+        }
+
+        try (RandomAccessFile file = new RandomAccessFile(new File(cacheFile.getPath().toUri()), "r")) {
+            file.seek(key.getOffset() - cacheFile.getOffset());
+            file.readFully(buffer, offset, key.getLength());
```

Comment:
Do we actually need to read `key.getLength()` or we only need to read `key.getLength() - offset`?

Reply:
`offset` is the start offset of the buffer. `length` is the actual data length it needs to read.